adoptingerlang.org

```diff
@@ -3444,44 +3446,74 @@ initContainers:
+connection to a specific port.
+
+A =livenessProbe= is not required to ensure restart of failed containers. If the
+process in the container crashes or stops for any reason, it will already be
+restarted. The probe should only be used for cases the process does not crash
+but for some reason has deadlocked and can not respond to requests. For this
+reason if a =livenessProbe= is included it should be kept extremely simple. It is
+more often than not safer to leave out the =livenessProbe= entirely and rely on
+the node crashes, metric based alerting and auto-scaling.
+
+One part of auto-scaling is based on the =readinessProbe=. The =readinessProbe=
+tells Kubernetes the =Pod= is ready to receive traffic. When a =Pod= becomes =unready=
+because a container is failing the =readinessProbe=, it is removed from the
+=Service= resulting in traffic being directed at the existing =Pods=, increasing
+their load, then a new =Pod= will be started.
+
+Not using the =livenessProbe= for this purpose means the containers are still
+available, since they aren't killed, and the reason for failing the check can be
+examined. This may mean attaching a shell if it isn't completely frozen or
+forcing a crash dump to be written. In a very resource constrained environment
+you may prefer to have the =livenessProbe= in order to have deadlocked containers
+and their =Pods= cleaned up right away, instead of waiting for manual intervention
+and the auto-scaler to scale down after load falls.
+
+Another use for the =readinessProbe= is if the application has to periodically do work
+where you don't want to be handling requests, or have to do some sort of
+maintenance, it can return a =503= status response for readiness causing it to be
+removed from the =Service= backends.
+
+Nothing special is needed to have a graceful shutdown, for any reason, whether
+it is during a rolling deploy or scale down. Erlang will call =init:stop()= when
+=SIGTERM= is sent by Kubernetes to signal shutdown, so each individual OTP
+Application should perform any necessary shutdown work in their =stop/1= function
+of the Application behaviour module as usual. In the case of =service_discovery=
+there is no shutdown work aside from trying to complete ongoing requests
+successfully. The Applications =service_discovery_http= and =service_discovery_grpc=
+can implement logic to delay the return from =stop= until active requests have
+been handled.
+
+By default the process can wait up to 30 seconds (configurable with the
+=terminationGracePeriodSeconds= =PodSpec= option) before Kubernetes sends a =SIGKILL=
+signal that will force the process to terminate.
+
+An important note about deployments and =readinessProbe= is the behaviour if a
+=readinessProbe= never passes. During a deployment if the container in the =Pod=
+never passes its =readinessProbe= then =Pods= from the previous version of the
+deployment that are =Ready= remain until they do.
```

Comment:
```suggestion
An important note about deployments and =readinessProbe=: 
During a deployment, if the container in the newly deployed =Pod=
never passes its =readinessProbe=, then =Pods= from the previous version of the
deployment that are =Ready= stick around until they do.
```

As a specific note, this is only true if you are using a `spec.strategy.type` of `RollingUpdate` rather than `Recreate`, and two options are in play here as well:

- `maxSurge`: The number of pods that can be created above the desired amount of pods during an update
- `maxUnavailable`: The number of pods that can be unavailable during the update process

K8s will happily kill up until `maxUnavailable` pods during a replacement before it starts hanging on the older ones because the new ones aren't around afaict.

Reply:
We may want to move that concern to the RollingDeployment section below.

Reply:
Updated.